tests/testapp/tests/test_controller_methods.py

```diff
@@ -0,0 +1,254 @@
+        new_store_record = Store.objects.get(id=self.fac1.id)
+
+        self.assertEqual(new_rmc.counter, new_store_record.last_saved_counter)
+        self.assertEqual(new_rmc.instance_id, new_store_record.last_saved_instance)
+
+    def test_update_rmc_for_existing_model(self):
+        Facility.objects.update(name='facility')
+        self.mc._serialize_into_store()
+
+        # there should only be 1 RecordMaxCounter for a specific instance_id and a specific model (unique_together)
+        self.assertEqual(RecordMaxCounter.objects.filter(instance_id=self.current_id.id, store_model_id=self.fac1.id).count(), 1)
+
+        new_rmc = RecordMaxCounter.objects.get(instance_id=self.current_id.id, store_model_id=self.fac1.id)
+        new_store_record = Store.objects.get(id=self.fac1.id)
+
+        self.assertEqual(self.old_rmc.counter + 1, new_rmc.counter)
+        self.assertEqual(new_rmc.counter, new_store_record.last_saved_counter)
+        self.assertEqual(new_rmc.instance_id, new_store_record.last_saved_instance)
+
+    def test_new_rmc_for_non_existent_model(self):
+        with mock.patch('platform.platform', return_value='Windows'):
+            (new_id, _) = InstanceIDModel.get_or_create_current_instance()
+
+        new_fac = FacilityModelFactory(name='college')
+        self.mc._serialize_into_store()
+        new_rmc = RecordMaxCounter.objects.get(instance_id=new_id.id, store_model_id=new_fac.id)
+        new_store_record = Store.objects.get(id=new_fac.id)
+
+        self.assertNotEqual(new_id.id, self.current_id.id)
+        self.assertEqual(new_store_record.last_saved_instance, new_rmc.instance_id)
+        self.assertEqual(new_store_record.last_saved_counter, new_rmc.counter)
+
+
+class DeserializationFromStoreIntoAppTestCase(TestCase):
+
+    def setUp(self):
+        DatabaseIDModel.objects.create()
+        (self.current_id, _) = InstanceIDModel.get_or_create_current_instance()
+        self.range = 10
+        self.mc = MorangoProfileController('facilitydata')
+        for i in range(self.range):
+            self.ident = uuid.uuid4().hex
+            StoreModelFacilityFactory(pk=self.ident, serialized=serialized_facility_factory(self.ident))
+
+    def test_dirty_store_records_are_deserialized(self):
+        self.assertFalse(Facility.objects.all())
+        self.mc._store_to_app()
+        self.assertEqual(len(Facility.objects.all()), self.range)
+
+    def test_clean_store_records_do_not_get_deserialized(self):
+        self.assertFalse(Facility.objects.all())
+        Store.objects.update(dirty_bit=False)
+        self.mc._store_to_app()
+        self.assertFalse(Facility.objects.all())
+
+    def test_deleted_models_do_not_get_deserialized(self):
+        Store.objects.update_or_create(defaults={'deleted': True}, id=self.ident)
+        self.mc._store_to_app()
+        with self.assertRaises(Facility.DoesNotExist):
+            Facility.objects.get(id=self.ident)
```

Comment:
Ah, interesting -- we *do* actually want models marked as deleted (and dirty) in the store to then lead to deletion in the app. This test case is compatible with that being the case, but doesn't test it directly.